### DISCUSSION TITLE
tactile_rl/

### DISCUSSION THREAD

**Discussion by giscus** [elle-miller/elle-miller.github.io#2] [Announcements]
# tactile_rl/

Enhancing Tactile-based Reinforcement Learning for Robotic Control

https://elle-miller.github.io/tactile_rl/

**Comment by elle-miller** [comment]
Ever read a paper and wished you could **directly ask the authors a question**? Here is your chance! If something was unclear in the paper, I would be super happy to help clarify, whether technical details (e.g., self-supervised objectives, POMDP formulation), or high-level decisions (like why we focused on sparse binary contacts). We also welcome general thoughts, comments and ideas. Have a nice day 😊

**Comment by reem-waseda** [comment]
hello :) first of all i would like to say this paper has been very pleasant to read and the supplementary materials are very accessible and easy to understand and i really appreciate that!
i have some questions about the methodology: 
1. you explored the idea of including a memory rollout in the process, and i assume that the training was done using on-policy experiences and then discarded after the policy has changed. however, towards the end of the paper you talked about the benefits of using off-policy data and im wondering then how the training was done using the memory and if you have tested using both on- and off-policy or just one method (and what the results were).
2. you talked about how pixels include very dense information that often overshadows the benefit of tactile sensors, and with the absence of pixels, it makes sense to turn to using methods like SSL in order to learn meaningful representations. however, im wondering if you tried using vision (in any sort of strategy) with this system and how it affects the learning process? it would also be interesting to see how a vision system compares to tactile in these tasks

thanks again!

**Reply by elle-miller** [reply]
thanks for the feedback and questions! :)

1. Yes, PPO training was done using the on-policy "RL" rollout which is then discarded. The "SSL/auxiliary" memory was a separate memory buffer, that contained the last $N_{rollouts}$ rollouts. So if $N_{rollouts}=4$, the memory buffer contained the most recent RL rollout (on-policy data), and the following three most recent rollouts (off-policy data). In the PPO optimisation loop, it is then just a matter of choosing where to get the auxiliary minibatches from (see code snippet below). In the paper, upto Figure 4 is just for the shared memory (ie $N_{rollouts}=1$). Figure 5 then shows what happens if I increase $N_{rollouts}$. While there was a negligible impact on Find/Bounce, the effect was significant for Baoding because I think a two-ball rotation spans more timesteps than a bounce or a contact, and the elongated memory was beneficial for dynamics learning.

```
# sample mini-batches from RL memory
 sampled_batches = self.memory.sample_all(names=self._tensors_names, mini_batches=self._mini_batches)

# auxiliary mini-batches either identical to RL memory or sampled from auxiliary memory
if self.ssl_task.use_same_memory:
     sampled_aux_batches = sampled_batches
else:
     sampled_aux_batches = self.ssl_task.memory.sample_all(mini_batches=self._mini_batches)
```
2. I would say training with pixels (whether RGB or tactile) slows down research turnaround for those with constrained compute, because it significantly restricts the number of parallelised environments the agent has access too plus the rendering slows down the simulation stepping. I included to this note to help others: I started this research using high-resolution tactile sensors, but it took so long to train that it was hard to make progress. It was only when I switched to simpler information that I could quickly iterate and make more findings. Thus my personal takeaway from this project is to start with simple observations, enable the agent to master those, and then incrementally scale up to more complex observations. I have not yet tried training from vision for these tasks, but it is included in the code so will try get around to it + post results at some point :) It will be possible but will require good camera angles, rendering settings, and different hyperparameters. I am also very curious how vision-based agents compare to tactile-based agents, and if vision + tactile < or > tactile!

**Comment by mxp1234** [comment]
Thank you for your inspiring work👏👏. I wonder if you have any open-source plans

**Reply by elle-miller** [reply]
Thanks so much! 💖 Yes everything is open-source on GitHub. See [roto](https://github.com/elle-miller/roto) for the environments, [multimodal_rl](https://github.com/elle-miller/multimodal_rl) for the RL+SSL agent training, and logs/checkpoints from the paper [here](https://github.com/elle-miller/roto_paper_results).

**Comment by elias0819** [comment]
Your work is really great and very inspiring! 
I am currently trying to implement the Baoding task with a different dexterous hand, but I am having trouble finding suitable target points. In many of my experiments, the two balls tend to stay around a certain position and keep collecting distance reward, instead of exploring a successful rotation behavior. I wanted to ask whether you have any experience or practical intuition for setting the target points in this task.
I also wanted to ask whether you have ever tried doing the Baoding task with the wrist fixed. In your experience, is it still possible to solve the task successfully without wrist motion? 
Thanks again for your time and excellent work.

**Comment by elle-miller** [comment]
Hi, thanks so much for your interest in this work! 

**Target points.** Yes, it took me a lot of trial-and-error to find good target points for the Shadow Hand. The euclidean distance from the ball to the target ball that is considered a "success" was also finnicky. If you send me an email (elle (dot) miller (at) ed.ac.uk) with details/screenshots of your current configuration happy to help and give suggestions.

**Failed runs.** As described in the paper, the performance was really stochastic in general for Baoding, it was a 50/50 split for success/failure with different seeds for the RL-only agent (why the return distributions are so wide). Two suggestions
1. While you are prototyping this new env, put "ground-truth" observations in, like the ball xyz positions and distances of each ball to current target. This task is super difficult to learn blind, so start with the easy case to make sure its possible with your new morphology.
2. Are you running the "RL only" version of the BaodingEnv (i.e. with no self-supervision)? Try training with `--agent_cfg forward_dynamics` for better performance (but will be slower).

 **Wrist motion.** I am currently working on redoing BaodingEnv for the Shadow Lite, Allegro and ORCA hand, so will have more insights soon I can share! For the Shadow Hand, you could test this by removing the 2x wrist joints from the actuated dofs. My gut feeling is yes still possible without wrist. If you look at the video, the wrist only moves side-to-side slightly (compare with how wrist moves in the Bounce video). It is the fingers doing the heavy lifting.
